### PR TITLE
Feature/wjiang/optimized writing node

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
The way how to choose a processor to write data is changed. Previously, a writing processor is chosen as long as it is idle. That may lead to a situation that some nodes are crowded but some others are idle.  Now, an idle processor is chosen from a node with the most idle processors. This is the first step to optimize the o-server. The next step would be using memory size as a criteria to choose writers.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: pfio/MultiGroupServer.F90 [hunk 1]
```diff
@@ -447,65 +447,103 @@ subroutine start_back(this, rc)
...
-                  endif
-               enddo
-
-               ! if all workers are busy, wait for one
-               if (idle_worker == 0) then 
-                  call MPI_recv( idle_worker, 1, MPI_INTEGER, &
-                      MPI_ANY_SOURCE, stag, this%back_comm, &
-                      MPI_STAT, ierr)
-               endif
+
```
Comment (tclune) [inline pfio/MultiGroupServer.F90:483]: This subroutine is getting long.   I encourage you to try to move some of this logic into well-named helper procedures to keep the logic in reasonably sized chunks.

**Review by tclune — CHANGES_REQUESTED** [review]
Complexity is just getting a bit high.  Time to refactor while it is fresh in your head.

**Comment by bena-nasa** [comment]
@weiyuan-jiang  Is there any danger here:
https://github.com/GEOS-ESM/MAPL/blob/feature/wjiang/optimized_writing_node/pfio/MultiGroupServer.F90#L486
since technically the "first" node has one less writer. I presume this works but it's getting late in the day. I'm trying to do something similar in my ESMF server prototype and was looking at your code when I saw this.

File: pfio/MultiGroupServer.F90 [hunk 2]
```diff
@@ -418,301 +419,352 @@ end subroutine receive_output_data
...
+                flag = .false.
+                call MPI_Iprobe( local_rank, stag, this%back_comm, flag, MPI_STAT, ierr)             
+                if (flag) then
+                  call MPI_recv(idle_worker, 1, MPI_INTEGER, &
+                                local_rank, stag, this%back_comm, &
+                                MPI_STAT, ierr)
+                  _ASSERT(local_rank == idle_worker, "local_rank and idle_worker should match")
+                  node_rank = this%node_ranks(local_rank)
+                  num_idlePEs(node_rank) = num_idlePEs(node_rank) + 1
+                  nth_writer = mod(local_rank, nwriter_per_node)
```
Comment (bena-nasa) [inline pfio/MultiGroupServer.F90:486]: See previous comment I made. Should have made it here.

File: pfio/MultiGroupServer.F90 (hunk 1 above)
Comment (weiyuan-jiang) [reply pfio/MultiGroupServer.F90]: I have re-organized the codes. And it seems more clear now

File: pfio/MultiGroupServer.F90 (hunk 2 above)
Comment (weiyuan-jiang) [reply pfio/MultiGroupServer.F90]: I re-organized the codes before I saw your comment. Can you comment again and let me what is the problem?

Comment (weiyuan-jiang) [reply pfio/MultiGroupServer.F90]: I think the fact that the first node has one processor less has been carefully taken care of all the time.

**Review by tclune — APPROVED** [review]
Definitely an improvement.   But there are still are some longish procedures.   

However, we want to get this released for benchmarking, so approving ...

**Review by tclune — APPROVED** [review]

**Pull request merged**